### ISSUE TITLE
[BUG] readlink /var/lib/docker/overlay2: invalid argument

### ISSUE THREAD

**Issue by Katawann** [docker/compose#10254]
### Description

When pulling a new version of a running docker, it fails to start the new one if the previous one has not been stopped properly:
```
Error response from daemon: readlink /var/lib/docker/overlay2: invalid argument
```

I found a lot of issues on this topic, but most of the time the solution was to `system prune -a` to clear the data which is not a solution in my use case. Not sure if there are some flag I could use to ensure it won't be block as it is right now

### Steps To Reproduce

1. Start a docker with docker-compose:
```
$ docker-compose -f docker-compose.hello1.yml up
[+] Running 5/5
 ⠿ hello_world Pulled                                             82.0s
   ⠿ a259ad5a4933 Pull complete                                   77.5s
   ⠿ e1753613a5b0 Pull complete                                   77.9s
   ⠿ 4f4fb700ef54 Pull complete                                   78.3s
   ⠿ e073d0d7e0b8 Pull complete                                   78.7s
[+] Running 2/2
 ⠿ Network aeros_default  Created                                  0.3s
 ⠿ Container hello_world  Created                                  4.5s
Attaching to hello_world
hello_world  | Hello 0 v1.0.0
hello_world  | Hello 1 v1.0.0
hello_world  | Hello 2 v1.0.0
...
```

2. Pull the new version of the docker and wait to be fully downloaded
```
$ docker-compose -f docker-compose.hello2.yml pull
[+] Running 5/5
 ⠿ hello_world Pulled                                          5.7s
   ⠿ a259ad5a4933 Already exists                               0.0s
   ⠿ 6ab636bd9b90 Pull complete                                1.5s
   ⠿ 4f4fb700ef54 Pull complete                                1.9s
   ⠿ 9d852e216d02 Pull complete 
```
3. Unplug the Raspberry from power while `hello_world` V1 is running
4. Plug again and once reboot, start the new pulled hello2 with:
```
$ docker-compose -f docker-compose.hello2.yml up
[+] Running 1/1
 ⠿ hello_world Pulled                                                                                   2.2s
Error response from daemon: readlink /var/lib/docker/overlay2: invalid argument
```
5. It should fail with the error above

Note:
  - If you press Ctrl+C (even twice to force) the new container start with **no error**
  - If the image is pulled while the other container is not running, there is **no error**
  - `docker-compose.hello1.yml` and `docker-compose.hello2.yml` contains a [basic loop script](https://github.com/rquackenbush/hello-world-loop) which print the version and are provided in the details section
  - The error actually also happen when using docker only:
```
  docker run aero41/hello-world:1.0.0
  docker pull aero41/hello-world:2.0.0
```
  and:
```
  $ docker run aero41/hello-world:2.0.0
Unable to find image 'aero41/hello-world:2.0.0' locally
2.0.0: Pulling from aero41/hello-world
Digest: sha256:758293899a851f4adf14e825c20cfdd777e8a75fccb0d451fd07ce9091a53413
Status: Downloaded newer image for aero41/hello-world:2.0.0
docker: Error response from daemon: error creating overlay mount to /var/lib/docker/overlay2/4e7acfa702a17b9c7328975f9ade291c51661d95b921f7ba013cfddd028d7121-init/merged: too many levels of symbolic links.
See 'docker run --help'.
```

  Error response from daemon: error creating overlay mount to /var/lib/docker/overlay2/4e7acfa702a17b9c7328975f9ade291c51661d95b921f7ba013cfddd028d7121-init/merged: too many levels of symbolic links.

### Compose Version

```Text
Docker Compose version v2.15.1
```

### Docker Environment

```Text
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.15.1
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 2
  Running: 1
  Paused: 0
  Stopped: 1
 Images: 3
 Server Version: 23.0.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 31aa4358a36870b21a992d3ad2bef29e1d693bec
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.10.92-v7l+
 Operating System: Raspbian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: armv7l
 CPUs: 4
 Total Memory: 1.828GiB
 Name: AGV216-9999
 ID: e4714198-ef4b-456e-a985-ed3d4e7f9b4f
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: aero41
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No memory limit support
WARNING: No swap limit support
```

### Anything else?

docker-compose.hello1.yml:
```
version: "3"
services:

  hello_world:
    image: aero41/hello-world:1.0.0
    container_name: hello_world
```

docker-compose.hello2.yml:
```
version: "3"
services:

  hello_world:
    image: aero41/hello-world:2.0.0
    container_name: hello_world
```

**Comment by milas** [comment]
The test case is working fine for me - I can pull the new image with the old one running, then issue a new `compose up` and everything works. This is the case if I kill the old one (to simulate a reboot) or not

> Unplug the Raspberry from power while hello_world V1 is running

It's very likely this is causing filesystem corruption, which is the root cause of errors like you're seeing

(I'd also monitor your Raspberry Pi's CPU/memory usage: you've only got 2GB of RAM, so might be having other issues depending on how much else is running & how much RAM it's using / how big the images you're pulling are, etc)

I'm gonna close this, as it does not appear to be a Compose-specific issue and is out of the scope of support we can provide here.

**Comment by Katawann** [comment]
Thank you for your feedback. I found a workaround. Probably not the best solution but for the sake of the next person having the same issue.

- I use `docker-compose -f docker-compose.hello2.yml create --pull=always` instead of `docker-compose -f docker-compose.hello2.yml pull`
- I gave different name for each image in docker-compose file:

docker-compose.hello1.yml:
```
version: "3"
services:

  hello_world_1:
    image: aero41/hello-world:1.0.0
    container_name: hello_world
```

docker-compose.hello2.yml:
```
version: "3"
services:

  hello_world_2:
    image: aero41/hello-world:2.0.0
    container_name: hello_world
```
